### PR TITLE
Migration of datatree/ops.py -> datatree_ops.py

### PR DESCRIPTION
**Pull request by flamingbear**
I considered wedging this into core/ops.py, but it didn't look like it fit there.

This is a basic lift and shift from datatree_/ops.py to core/datatree_ops.py

I did fix the document addendum injection and added a couple of tests.

- [x]  Contributes to migration step for miscellaneous modules in Track merging datatree into xarray #8572
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

**Comment by flamingbear** [comment]
Here's a sample of a doc page generated with these updates.  I will point to the RTD when that finishes.

Screenshot 2024-04-26 at 2 07 38 PM

**Comment by flamingbear** [comment]
Ok. The test build for docs is available to examine here https://xray--8977.org.readthedocs.build/en/8977/generated/xarray.DataTree.expand_dims.html#xarray.DataTree.expand_dims

File: xarray/tests/test_datatree.py [hunk 1]
```diff
@@ -824,3 +826,79 @@ def test_tree(self, create_test_datatree):
         expected = create_test_datatree(modify=lambda ds: np.sin(ds))
         result_tree = np.sin(dt)
         assert_equal(result_tree, expected)
+
+
+class TestDocInsertion:
```
Comment (owenlittlejohns) [inline xarray/tests/test_datatree.py:831]: Nice tests.

File: xarray/core/datatree_ops.py [hunk 2]
```diff
@@ -208,16 +207,62 @@ def method_name(self, *args, **kwargs):
...
+    """Insert addendum after first paragraph or at the end of the docstring.
+
+    There are a number of Dataset's functions that are wrapped. These come from
+    Dataset directly as well as the mixins: DataWithCoords, DatasetAggregations, and DatasetOpsMixin.
+
+    The majority of the docstrings fall into a parseable pattern. Those that
+    don't, just have the addendum appeneded after. None values are returned.
+
+    """
+    if docstring is None:
```
Comment (owenlittlejohns) [inline xarray/core/datatree_ops.py:228]: Nit: I think `docstring` is guarded against being `None` from the call to this function being wrapped in `if orig_method_docstring is not None`. It doesn't look like this function is called elsewhere, but I might be missing something.

File: xarray/core/datatree_ops.py [hunk 3]
```diff
@@ -208,16 +207,62 @@ def method_name(self, *args, **kwargs):
...
-            setattr(target_cls_dict[method_name], "__doc__", orig_method_docstring)
+
+            if orig_method_docstring is not None:
+                new_method_docstring = insert_doc_addendum(
+                    orig_method_docstring, _MAPPED_DOCSTRING_ADDENDUM
+                )
+                setattr(target_cls_dict[method_name], "__doc__", new_method_docstring)
+
+
+def insert_doc_addendum(docstring, addendum):
```
Comment (owenlittlejohns) [inline xarray/core/datatree_ops.py:218]: Nit: Is it overkill to suggest type hints for this function?

File: xarray/tests/test_datatree.py [hunk 4]
```diff
@@ -824,3 +826,79 @@ def test_tree(self, create_test_datatree):
...
+    def test_one_liner(self):
+        mixin_doc = "Same as abs(a)."
+
+        expected_doc = dedent(
+            """\
+            Same as abs(a).
+
+            This method was copied from xarray.Dataset, but has been altered to call the
+                method on the Datasets stored in every node of the subtree. See the
+                `map_over_subtree` function for more details."""
```
Comment (owenlittlejohns) [inline xarray/tests/test_datatree.py:896]: So, when this message is just appended on the end of a single paragraph/line docstring, it isn't a note? I don't have strong opinions, but just highlighting the difference and asking if it should still be a note.

**Review by owenlittlejohns — COMMENTED** [review]
Makes sense and the compiled documentation looks good in the RTD build. I think I only had one question really. Nice work!

File: xarray/core/datatree_ops.py (hunk 2 above)
Comment (flamingbear) [reply xarray/core/datatree_ops.py:228]: It is, and I couldn't decide if I should take it out above or leave it here, so it's belts and braces.

File: xarray/core/datatree_ops.py (hunk 3 above)
Comment (flamingbear) [reply xarray/core/datatree_ops.py:218]: probably not

File: xarray/tests/test_datatree.py (hunk 4 above)
Comment (flamingbear) [reply xarray/tests/test_datatree.py:896]: Yes, I already don't know if I did this on purpose or not.  All of the single
line docstrings from from DatasetOpsMixin dunder methods,
e.g. 
```
DatasetOpsMixin.__ifloordiv__ 'Same as a //= b.'
DatasetOpsMixin.__rfloordiv__ 'Same as a // b.'
DatasetOpsMixin.__rmod__ 'Same as a % b.'
etc.
```  

I don't see in the RTD docs if those are not built, or if I just didn't insert
them. I can't find any of those on the Dataset objects either.  I think I was
thinking you'd most likely see the doc in a repl, but who knows why that's what
I was thinking..

File: xarray/core/datatree_ops.py (hunk 3 above)
Comment (flamingbear) [reply xarray/core/datatree_ops.py]: 696cd3e

**Comment by TomNicholas** [comment]
This looks great, thank you @flamingbear ! Happy to merge?

**Comment by flamingbear** [comment]
> Happy to merge?

I am.  I might take on docs next.

**Pull request merged**